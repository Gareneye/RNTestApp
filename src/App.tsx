/**
 * App Initial Component
 */

import { NavigationContainer } from '@react-navigation/native';
import React from 'react';
import { RootStackNavigator } from './navigators/RootStackNavigator';
import { navigationRef } from './utilities/Navigator';

export default function App() {
  return (
    <NavigationContainer ref={navigationRef}>
      <RootStackNavigator />
    </NavigationContainer>
  );
}
