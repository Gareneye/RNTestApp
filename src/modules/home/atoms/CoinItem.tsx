import React from 'react';
import { StyleSheet, Text, Image, View } from 'react-native';
import { Ticker } from 'models/Ticker';
import { Fonts, Dimens, Colors } from 'resources';
import { SvgUri } from 'react-native-svg';
import { Quote } from 'models/Quote';

interface Props {
  item: Ticker;
}

export const MAX_TITLE_LENGTH = 12;
export const COIN_ITEM_MIN_HEIGHT = 45;

export const CoinItem = (props: Props) => {
  const ticker = props.item;
  const imageUrl = `https://static2.coinpaprika.com/coin/${ticker.id}/logo-thumb.png`;
  const chartUrl = `https://graphs2.coinpaprika.com/currency/chart/${ticker.id}/7d/chart.svg`;

  const tempQuote: Quote | undefined = ticker.quotes['USD'];
  const value = tempQuote.price.toFixed(2);
  const percentChange = tempQuote.percent_change_7d;

  return (
    <View style={styles.row}>
      <View style={styles.col}>
        <View style={styles.headerRow}>
          <Image source={{ uri: imageUrl }} style={styles.icon} />
          <Text style={styles.title}>{ticker.symbol}</Text>
        </View>
        <View style={[styles.headerRow, styles.paddedRow]}>
          <Text style={styles.rank}>{ticker.rank}</Text>
          <Text style={styles.desc}>
            {ticker.name.slice(0, MAX_TITLE_LENGTH)}
            {ticker.name.length > MAX_TITLE_LENGTH ? '...' : null}
          </Text>
        </View>
      </View>

      <View style={styles.col}>
        <View style={styles.row}>
          <View style={[styles.col, styles.chartWrapper]}>
            <SvgUri width={100} height={30} uri={chartUrl} />
          </View>

          <View style={[styles.col, styles.lastCol, styles.pricesWrapper]}>
            <Text style={styles.price}>{`$ ${value}`}</Text>
            <View style={[styles.headerRow, styles.paddedRow]}>
              <View
                style={[
                  styles.changeLabel,
                  percentChange > 0
                    ? styles.changeLabelUp
                    : styles.changeLabelDown,
                ]}>
                <Text
                  style={styles.changeLabelText}>{`${percentChange}%`}</Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  row: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    minHeight: COIN_ITEM_MIN_HEIGHT,
  },
  col: {
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  lastCol: {
    alignItems: 'flex-end',
  },
  headerRow: {
    flexDirection: 'row',
  },
  paddedRow: {
    marginTop: Dimens.space.s,
  },
  pricesWrapper: {
    minWidth: 80,
  },
  chartWrapper: {
    justifyContent: 'flex-end',
    marginRight: Dimens.space.xl,
  },
  icon: {
    width: 20,
    height: 20,
    marginRight: Dimens.space.s,
  },
  title: {
    ...Fonts.label15,
  },
  rank: {
    ...Fonts.caption13,
    backgroundColor: Colors.border,
    lineHeight: 20,
    minWidth: 20,
    height: 20,
    textAlign: 'center',
    paddingHorizontal: Dimens.space.xs,
    marginRight: Dimens.space.s,
    color: Colors.desc,
  },
  desc: {
    ...Fonts.body15,
    color: Colors.desc,
  },
  changeLabel: {
    minWidth: 20,
    height: 24,
    alignItems: 'flex-end',
    paddingHorizontal: Dimens.space.s,
    borderRadius: 5,
  },
  changeLabelText: {
    ...Fonts.label15,
    lineHeight: 24,
    color: 'white',
  },
  changeLabelUp: {
    backgroundColor: 'green',
  },
  changeLabelDown: {
    backgroundColor: 'red',
  },
  price: {
    ...Fonts.body15,
  },
});
