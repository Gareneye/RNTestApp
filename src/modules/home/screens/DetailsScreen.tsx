import React from 'react';
import { SafeAreaView, StatusBar, StyleSheet, Text, View } from 'react-native';
import { Colors } from 'react-native/Libraries/NewAppScreen';

export const DetailsScreen = () => {
  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView style={styles.wrapper}>
        <View style={styles.body}>
          <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Details</Text>
          </View>
        </View>
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
  },
  scrollView: {
    backgroundColor: Colors.lighter,
  },
  engine: {
    position: 'absolute',
    right: 0,
  },
  body: {
    backgroundColor: Colors.white,
  },
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
    color: Colors.black,
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
    color: Colors.dark,
  },
  highlight: {
    fontWeight: '700',
  },
  footer: {
    color: Colors.dark,
    fontSize: 12,
    fontWeight: '600',
    padding: 4,
    paddingRight: 12,
    textAlign: 'right',
  },
});
