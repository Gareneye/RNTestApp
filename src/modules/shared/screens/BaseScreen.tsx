import { NavigationProp } from '@react-navigation/core';
export interface IBaseScreenProps {
  navigation: NavigationProp<{}>;
}
