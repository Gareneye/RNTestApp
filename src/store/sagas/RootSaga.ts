import { select, spawn, take, fork, call, put, takeLatest, takeLeading, takeEvery } from '@redux-saga/core/effects';
import { ActionId } from 'store/ActionId';
import ApiService from 'utilities/ApiService';
import { fetchTickerSuccessAction, fetchTickerErrorAction, fetchTickerStartAction, sortTickersDoneAction } from 'store/Actions';
import { RootState } from 'store/RootState';
import { TickerListSorting, sortTickers } from 'helpers/TickersSortingStrategies';
import { Ticker } from 'models/Ticker';

export function* rootSaga() {
  yield spawn(watchTickersSaga);
  yield spawn(watchListSaga);
  yield spawn(watchTickersSortingSaga);
}

export function* watchListSaga() {
  yield take(ActionId.LIST_APPEARED_FIRST_TIME);
  yield put(fetchTickerStartAction());
}

export function* watchTickersSortingSaga() {
  while (true) {
    const action: { payload: { sortBy: TickerListSorting }} = yield take(ActionId.SORT_TICKERS_START);
    const store: RootState = yield select();
    const tickers = store.tickersBranch.tickers;
    yield spawn(sortTickersSaga, tickers, action.payload.sortBy);
  }
}

export function* sortTickersSaga(tickers: Ticker[], sortBy:TickerListSorting ) {
  const sorted = sortTickers(tickers, sortBy );
  yield put(sortTickersDoneAction(sorted))
}

export function* watchTickersSaga() {
  while (true) {
    yield take(ActionId.FETCH_TICKER_START);
    yield call(fetchTickersSaga);
  }
}

export function* fetchTickersSaga() {
  try {
    const tickers = yield call(ApiService.getTickersForAllCoins);
    yield put(fetchTickerSuccessAction(tickers))
    return tickers
  } catch(_error) {
    yield put(fetchTickerErrorAction())
  }
}