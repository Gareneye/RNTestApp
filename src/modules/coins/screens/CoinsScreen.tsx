import { BaseScreenProps } from 'modules/shared/screens/BaseScreen';
import React, { useEffect } from 'react';
import {
  SafeAreaView,
  StatusBar,
  StyleSheet,
  View,
  FlatList,
  Text,
  Dimensions,
  TouchableOpacity,
} from 'react-native';
import { connect } from 'react-redux';

import { RootState } from 'store/RootState';
import { Ticker } from 'models/Ticker';
import { CoinItem, COIN_ITEM_MIN_HEIGHT } from '../atoms/CoinItem';
import { Colors, Dimens, Fonts } from 'resources';
import {
  listAppearedFirstTimeAction,
  sortTickersStartAction,
} from 'store/Actions';
import { TickerListSorting } from 'helpers/TickersSortingStrategies';

const connector = connect(
  ({ tickersBranch }: RootState) => ({
    sorted: tickersBranch.sorted,
    isLoading: tickersBranch.isLoading,
    sortBy: tickersBranch.sortBy
  }),
  {
    initAction: () => listAppearedFirstTimeAction(),
    filterTickers: (sortBy: TickerListSorting) =>
      sortTickersStartAction(sortBy),
  },
);

type Props = {};

// Flat List Settings
const SEPARATOR_HEIGHT = Dimens.space.m;
const ROWS_ON_SCREEN = Math.ceil(
  Dimensions.get('screen').height /
    (COIN_ITEM_MIN_HEIGHT + 2 * SEPARATOR_HEIGHT),
);

const ItemSeparator = () => <View style={styles.separator} />;

export const CoinsScreen = connector(
  (props: Props & BaseScreenProps<typeof connector>) => {
    const renderItem = ({ item }: { item: Ticker }) => <CoinItem item={item} />;
    const keyExtractor = (item: Ticker) => item.id;
    const renderedScreensAmount = 17; // Default is 21

    useEffect(() => {
      props.initAction();
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return (
      <>
        <StatusBar barStyle="dark-content" />
        <SafeAreaView style={styles.wrapper}>
          {/* Todo: Refactor this */}
          <View style={{flexDirection: 'row'}}>
            <TouchableOpacity
              onPress={() => {
                props.filterTickers(TickerListSorting.LOWER_PRICE);
              }}>
              <Text style={{ ...Fonts.caption13, height: 30, lineHeight: 30, paddingHorizontal: 15, borderWidth: 1, margin: 20, backgroundColor: props.sortBy === TickerListSorting.LOWER_PRICE ? 'grey' : 'white'  }}>
                {'Sort by lower price'}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => {
                props.filterTickers(TickerListSorting.HIGHER_PRICE);
              }}>
              <Text style={{ ...Fonts.caption13, height: 30, lineHeight: 30, paddingHorizontal: 15, borderWidth: 1, margin: 20,backgroundColor: props.sortBy === TickerListSorting.HIGHER_PRICE ? 'grey' : 'white' }}>
                {'Sort by higher price'}
              </Text>
            </TouchableOpacity>
          </View>

          {props.isLoading ? (
            <Text style={{ paddingHorizontal: 20 }}>{'Loading...'}</Text>
          ) : (
            <FlatList
              contentContainerStyle={{ marginHorizontal: 20 }}
              data={props.sorted}
              renderItem={renderItem}
              keyExtractor={keyExtractor}
              ItemSeparatorComponent={ItemSeparator}
              windowSize={renderedScreensAmount}
              removeClippedSubviews={true}
              initialNumToRender={ROWS_ON_SCREEN}
              maxToRenderPerBatch={Math.min(10, ROWS_ON_SCREEN)}
              updateCellsBatchingPeriod={25}
            />
          )}
        </SafeAreaView>
      </>
    );
  },
);

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    backgroundColor: 'white',
  },
  separator: {
    flex: 1,
    height: 1,
    backgroundColor: Colors.separator,
    marginVertical: SEPARATOR_HEIGHT,
  },
});
