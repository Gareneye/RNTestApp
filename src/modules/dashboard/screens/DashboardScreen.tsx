import { BaseScreenProps } from 'modules/shared/screens/BaseScreen';
import React from 'react';
import { SafeAreaView, StatusBar, Text } from 'react-native';
import { connect } from 'react-redux';

import { RootState } from 'store/RootState';

const connector = connect(({}: RootState) => ({}), {});

type Props = {};

export const DashboardScreen = connector(
  (_props: Props & BaseScreenProps<typeof connector>) => {
    return (
      <>
        <StatusBar barStyle="dark-content" />
        <SafeAreaView>
          <Text>{'Dashboard screen'}</Text>
        </SafeAreaView>
      </>
    );
  },
);
