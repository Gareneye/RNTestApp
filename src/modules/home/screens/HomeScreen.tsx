import { IBaseScreenProps } from 'modules/shared/screens/BaseScreen';
import React from 'react';
import {
  SafeAreaView,
  ScrollView,
  StatusBar,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { Colors } from 'react-native/Libraries/NewAppScreen';
import { Navigator, RouteName } from 'utilities/Navigator';
import { Fonts } from 'resources/Fonts';

interface IProps extends IBaseScreenProps {}

export const HomeScreen = (_props: IProps) => {
  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView style={styles.wrapper}>
        <ScrollView
          contentInsetAdjustmentBehavior="automatic"
          style={styles.scrollView}>
          <View style={styles.body}>
            <View style={styles.sectionContainer}>
              <Text style={styles.sectionTitle}>Hey guys!</Text>
              <TouchableOpacity
                onPress={() => Navigator.navigate(RouteName.DetailsScreen)}>
                <Text style={styles.sectionDescription}>
                  Welcome in
                  <Text style={styles.highlight}> Pannoire Starter</Text>
                </Text>
              </TouchableOpacity>
            </View>
            <View style={styles.sectionContainer}>
              <Text style={styles.sectionTitle}>My contact:</Text>
              <Text style={styles.sectionDescription}>blogokodzie.pl</Text>
              <Text style={styles.sectionDescription}>blogokodzie.pl</Text>
            </View>
          </View>
        </ScrollView>
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
  },
  scrollView: {
    backgroundColor: Colors.lighter,
  },
  engine: {
    position: 'absolute',
    right: 0,
  },
  body: {
    backgroundColor: Colors.white,
  },
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    ...Fonts.body15,
    fontSize: 24,
    fontWeight: '600',
    color: Colors.black,
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
    color: Colors.dark,
  },
  highlight: {
    fontWeight: '700',
  },
  footer: {
    color: Colors.dark,
    fontSize: 12,
    fontWeight: '600',
    padding: 4,
    paddingRight: 12,
    textAlign: 'right',
  },
});
