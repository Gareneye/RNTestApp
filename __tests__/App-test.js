/**
 * @format
 */

import React from 'react';
import 'react-native';
import renderer, { act, test } from 'react-test-renderer'; // Note: test renderer must be required after react-native.
import App from '../src/App';
import { CoinItem } from 'modules/coins/atoms/CoinItem';
import { CoinsScreen } from 'modules/coins/screens/CoinsScreen';
import { cloneableGenerator } from '@redux-saga/testing-utils';
import { watchTickersSortingSaga } from 'store/sagas/RootSaga';

// Ticker mock
const ticker = {
  quotes: {
    USD: {
      price: 36650.73287245,
      volume_24h: 876.10205064,
      volume_24h_change_24h: 131.7,
      market_cap: 1539330,
      market_cap_change_24h: 8.46,
      percent_change_15m: -0.08,
      percent_change_30m: 2.56,
      percent_change_1h: 0.1,
      percent_change_6h: 8.61,
      percent_change_12h: 7.01,
      percent_change_24h: 8.46,
      percent_change_7d: 2.31,
      percent_change_30d: 22.6,
      percent_change_1y: 78.04,
      ath_price: 1146320,
      ath_date: '2014-01-14T05:47:00Z',
      percent_from_price_ath: -96.8
    }
  },
  id: '42-42-coin',
  name: '42-coin',
  symbol: '42',
  rank: 1152,
  circulating_supply: 42,
  total_supply: 42,
  max_supply: 0,
  beta_value: 0.91836,
  last_updated: '2020-08-20T14:43:35Z'
}

// 1. Render tests
it('app works', async () => {
  await act(async () => {
    renderer.create(<App />);
   })
});

it('CoinItem renders correctly',  () => {
  const tree = renderer.create(<CoinItem item={ticker} />).toJSON();
  expect(tree).toMatchSnapshot();
});

// 2. Saga tests
it('sortingSaga', assert => {
  const gen = cloneableGenerator(watchTickersSortingSaga)();
  // gen.next(); 
});


// 3. Sorting strategy test