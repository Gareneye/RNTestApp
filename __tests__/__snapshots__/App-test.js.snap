// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinItem renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "minHeight": 45,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://static2.coinpaprika.com/coin/42-42-coin/logo-thumb.png",
          }
        }
        style={
          Object {
            "height": 20,
            "marginRight": 8,
            "width": 20,
          }
        }
      />
      <Text
        style={
          Object {
            "fontFamily": "Lato-Semibold",
            "fontSize": 15,
            "lineHeight": 20,
          }
        }
      >
        42
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "marginTop": 8,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#D5DAD5",
            "color": "grey",
            "fontFamily": "Lato-Semibold",
            "fontSize": 13,
            "height": 20,
            "lineHeight": 20,
            "marginRight": 8,
            "minWidth": 20,
            "paddingHorizontal": 4,
            "textAlign": "center",
          }
        }
      >
        1152
      </Text>
      <Text
        style={
          Object {
            "color": "grey",
            "fontFamily": "Lato-Regular",
            "fontSize": 15,
            "lineHeight": 20,
          }
        }
      >
        42-coin
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 45,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "column",
              "justifyContent": "space-between",
            },
            Object {
              "justifyContent": "flex-end",
              "marginRight": 16,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "flexDirection": "column",
              "justifyContent": "space-between",
            },
            Object {
              "alignItems": "flex-end",
            },
            Object {
              "minWidth": 80,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontFamily": "Lato-Regular",
              "fontSize": 15,
              "lineHeight": 20,
            }
          }
        >
          $ 36650.73
        </Text>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "marginTop": 8,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-end",
                  "borderRadius": 5,
                  "height": 24,
                  "minWidth": 20,
                  "paddingHorizontal": 8,
                },
                Object {
                  "backgroundColor": "green",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontFamily": "Lato-Semibold",
                  "fontSize": 15,
                  "lineHeight": 24,
                }
              }
            >
              2.31%
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
